bundles/org.openhab.binding.caddx/src/main/java/org/openhab/binding/caddx/internal/CaddxCommunicator.java

```diff
@@ -0,0 +1,559 @@
+/**
+ * Copyright (c) 2010-2020 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.caddx.internal;
+
+import java.io.IOException;
+import java.io.InputStream;
+import java.io.OutputStream;
+import java.util.ArrayList;
+import java.util.TooManyListenersException;
+import java.util.concurrent.Exchanger;
+import java.util.concurrent.LinkedBlockingDeque;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.TimeoutException;
+import java.util.stream.IntStream;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.io.transport.serial.PortInUseException;
+import org.eclipse.smarthome.io.transport.serial.SerialPort;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEvent;
+import org.eclipse.smarthome.io.transport.serial.SerialPortEventListener;
+import org.eclipse.smarthome.io.transport.serial.SerialPortIdentifier;
+import org.eclipse.smarthome.io.transport.serial.SerialPortManager;
+import org.eclipse.smarthome.io.transport.serial.UnsupportedCommOperationException;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+/**
+ * The {@link CaddxCommunicator} is responsible for the asynchronous serial communication
+ *
+ * @author Georgios Moutsos - Initial contribution
+ */
+@NonNullByDefault
+public class CaddxCommunicator implements Runnable, SerialPortEventListener {
+    private final Logger logger = LoggerFactory.getLogger(CaddxCommunicator.class);
+
+    private final ArrayList<CaddxPanelListener> listenerQueue = new ArrayList<>();
+
+    private Thread thread;
+    private final LinkedBlockingDeque<CaddxMessage> messages = new LinkedBlockingDeque<>();
+
+    private CaddxProtocol protocol;
+    private String serialPortName;
+    private int baudRate;
+    private final SerialPortManager portManager;
+    private SerialPort serialPort;
+    private @Nullable InputStream in;
+    private @Nullable OutputStream out;
+    Exchanger<CaddxMessage> exchanger = new Exchanger<>();
```

Comment:
Could you tell me what you are using this for?

Reply:
A message can originate from the panel or from the binding.
Every message that is sent is acknowledged by the other party, either with a specific response message or with an ACK/NAK/Reject message (if a specific response message does not exist).
The window for the acknowledgement is 3 sec. If no acknowledgement is received, the message is resent.

The communicator uses two threads. One for sending and one for receiving.
The exchanger is used to send the received message from the Receiver thread to the Sender thread. The Sender thread checks if it is an expected response.

If it is not, then it means that the binding and the panel try concurrently to send a message to each other. In that case the Sender message is put again in front of the outgoing queue and on the next loop we skip sending of the message (to give priority to the panel) and we receive the originating panel message which is then acknowledged by the binding.

Reply:
The SynchronousQueue is exactly what is needed at this point. 
Didn't noticed this class when going through the Java reference manual. I will replace the Exchanger.